internal/rpc/core/events.go

```diff
@@ -126,3 +129,131 @@ func (env *Environment) UnsubscribeAll(ctx context.Context) (*coretypes.ResultUn
 	}
 	return &coretypes.ResultUnsubscribe{}, nil
 }
+
+// Events applies a query to the event log. If an event log is not enabled,
+// Events reports an error. Otherwise, it filters the current contents of the
+// log to return matching events.
+//
+// Events returns up to maxItems of the newest eligible event items. An item is
+// eligible if it is older than before (or before is zero), it is newer than
+// after (or after is zero), and its data matches the filter. A nil filter
+// matches all event data.
+//
+// If before is zero and no eligible event items are available, Events waits
+// for up to waitTime for a matching item to become available. The wait is
+// terminated early if ctx ends.
+//
+// If maxItems ≤ 0, a default positive number of events is chosen. The values
+// of maxItems and waitTime may be capped to sensible internal maxima without
+// reporting an error to the caller.
+func (env *Environment) Events(ctx context.Context,
+	filter *coretypes.EventFilter,
+	maxItems int,
+	before, after cursor.Cursor,
```

Comment:
do we need to support before and after? Seems like it'd be fine to just do "after", but perhaps I'm missing something

Reply:
We need both to paginate.

In normal operation, `before` == "", `after` is the latest thing you've seen (initially "") and you get whatever is at the head of the log. Then there are two cases:

1. You got everything you requested (or there were no results). Subsequent requests set `after` to skip past all the stuff you've already seen, and `before` == "".
2. You got some, but there's more. Subsequent requests leave `after` where it was, and set `before` to the oldest result in that group to get the next page. This continues until there are no more results, then you go back to normal operation.

This iteration model is more or less how a tweet stream works—indices into the stream are stable despite the head moving forward in time, and you control both how much you get and when to advance.

There are other ways you could get at the same effect, but I think this is probably the simplest (fewest moving parts). Another nice effect is it gives you an easy way to tell whether you lost some data on the client side.

Reply:
If this is the case, then is it possible to embed the before and after into the cursor object itself? If you can't use the cursor without having two of them? (I think I see the benefit to having this kind of cursor, though my existing mental model for cursors is different)

Reply:
> If this is the case, then is it possible to embed the before and after into the cursor object itself? If you can't use the cursor without having two of them? (I think I see the benefit to having this kind of cursor, though my existing mental model for cursors is different)

We could have an object that carries both, but the caller will still want to vary them both independently.   So putting them on a single data structure doesn't really reduce the variability, just how it's encoded.

As for the parameters: I'd really rather not have positional parameters here at all, but the way the RPC mechanism decodes requests it's tedious to pass in structured data. If we put a structure here, we either have to give it a string encoding or it won't be usable from the GET endpoint.  (Inventing a combined string encoding is definitely doable, but I think it would hurt UX)

Reply:
Suppose when there are multiple pages, the service returns a "combined" cursor that gets the "next page" and nothing else. Now the caller _has_ to read all the pages, even if they don't want to. Either that, or they need to know how to construct their own "combined" cursor. 

Neither of those really feels right to me—the value is just an opaque anchor in a time series, and making the client reason about paired vs. unpaired variants doesn't appear to simplify the experience. 

Or am I missing some aspect of your suggestion?

Reply:
yeah, I think I get it now. I had in my mind that every _event_ would have some kind of ID (instead we use the timestamp) and that the cursor itself could hold the ID of the end of the batch, but this doesn't quite make sense, but I think it's alright. 

I guess it's possible that two events could have the same timestamp, and you could end up with a duplicate event if you got very un/lucky.

Reply:
Yeah, I think a duplicate timestamp is actually quite unlikely, but that is why the sequence number is in there, just to ensure they're distinct. (Now to get a collision we'd have to return the same time stamp > 65K times consecutively, and I think that safe to ignore)

Even if we did (hypothetically) get a collision the worst that could happen is a client doesn't see all the items that share the duplicated cursor. Not ideal but not catastrophic either.